crates/compose/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+///
+/// TODO: describe this process (resolving dependency names to instantiation arguments, enforcement of configuration inheritance, etc.)
+pub async fn compose<'a, L: ComponentLoader>(
+    loader: &'a L,
+    component: &LockedComponent,
+) -> anyhow::Result<Vec<u8>> {
+    Composer::new(loader).compose(component).await
+}
+
+/// This trait is used to load component source code from a locked component source across various embdeddings.
+#[async_trait::async_trait]
+pub trait ComponentLoader {
+    async fn load_component(
+        &self,
+        source: &locked::LockedComponentSource,
+    ) -> anyhow::Result<Vec<u8>>;
+}
+
+struct Composer<'a, L> {
+    graph: CompositionGraph,
+    loader: &'a L,
+}
+
+impl<'a, L: ComponentLoader> Composer<'a, L> {
+    async fn compose(mut self, component: &LockedComponent) -> anyhow::Result<Vec<u8>> {
+        let source = self.loader.load_component(&component.source).await?;
+
+        if component.dependencies.is_empty() {
+            return Ok(source);
+        }
+
+        let package = Package::from_bytes(&component.id, None, source, self.graph.types_mut())?;
+
+        let world_id = package.ty();
+        let package_id = self.graph.register_package(package)?;
+        let instantiation_id = self.graph.instantiate(package_id);
+
+        let prepared = self.prepare_dependencies(world_id, component).await?;
+
+        let arguments = self
+            .build_instantiation_arguments(world_id, prepared)
+            .await?;
+
+        for (argument_name, argument) in arguments {
+            self.graph
+                .set_instantiation_argument(instantiation_id, &argument_name, argument)?;
+        }
+
+        // Export all exports from the root component
+        for export_name in self.graph.types()[world_id]
+            .exports
+            .keys()
+            .cloned()
+            .collect::<Vec<_>>()
+        {
+            let export_id = self
+                .graph
+                .alias_instance_export(instantiation_id, &export_name)?;
+
+            self.graph.export(export_id, &export_name)?;
+        }
+
+        self.graph
+            .encode(Default::default())
+            .with_context(|| "failed to encode composition graph")
+            .map(Into::into)
+    }
+
+    fn new(loader: &'a L) -> Self {
+        Self {
+            graph: CompositionGraph::new(),
+            loader,
+        }
+    }
+
+    async fn prepare_dependencies(
+        &mut self,
+        world_id: WorldId,
+        component: &LockedComponent,
+    ) -> anyhow::Result<IndexMap<String, DependencyInfo>> {
+        let mut expanded = IndexMap::new();
+
+        let mut imports = self.graph.types()[world_id].imports.clone();
+
+        // TODO: collect duplicates and return
+
+        for (dependency_name, dependency) in component.dependencies.iter() {
+            let import_keys = imports.keys().cloned().collect::<Vec<_>>();
```

Comment:
Shouldn't we be collecting this outside the iteration of dependencies? Otherwise we reset the list of imports on each iteration.

Reply:
We reset the imports here because we are consuming them from `imports` as we match against dependency names. Since after each match, the set of imports is potentially smaller so we don't want to walk to full set of import keys on each iteration. This is just a slight optimization to keep the complexity best-case O(m*log(n)), worst-case O(n*m).

Reply:
Ah ok makes sense. Can we document that with a comment?

Reply:
Updated, PTAL